subsys/bluetooth/host/settings.c

```diff
@@ -298,6 +301,124 @@ int bt_settings_init(void)
 	return 0;
 }
 
+static const char * const bt_settings_keys[] = {
+	"sc",
+	"cf",
+	"ccc",
+	"hash",
+	"name",
+	"id",
+	"irk",
+	"link_key",
+	"keys",
+	"mesh",
+};
+
+static int bt_settings_load_direct_cleanup(const char *key, size_t len, settings_read_cb read_cb,
+					   void *cb_arg, void *param)
+{
+	int err;
+	int next_len;
+	int subsys_len;
+	unsigned long id;
+	const char *next;
+	const char *id_str;
+	bool key_is_supported;
+	char full_key[BT_SETTINGS_KEY_MAX];
+
+	bt_addr_le_t *id_addr = (bt_addr_le_t *)param;
+
+	subsys_len = settings_name_next(key, &next);
+
+	LOG_DBG("subsys: %.*s", subsys_len, key);
+
+	key_is_supported = false;
+
+	for (size_t i = 0; i < ARRAY_SIZE(bt_settings_keys); i++) {
+		if (!strncmp(key, bt_settings_keys[i], subsys_len)) {
+			key_is_supported = true;
+		}
+	}
+
+	if (key_is_supported) {
+		if (!next || !strncmp(key, "mesh", subsys_len)) {
+			/* key exist but does not need to be checked for
+			 * leftover because it is unique and not linked to ID.
+			 * We return and continue to look for next keys.
+			 */
+			return 0;
+		}
+
+		next_len = settings_name_next(next, &id_str);
+
+		if (!id_str) {
+			id = BT_ID_DEFAULT;
+		} else {
+			id = strtoul(id_str, NULL, 10);
+
+			if (id >= CONFIG_BT_ID_MAX) {
+				LOG_ERR("Invalid local identity (%lu)", id);
+
```

Comment:
Should we delete when this happens? RFC

Reply:
I am not sure that we want to delete in this case. Maybe the user misconfigured its device and will update the configuration when it see the error.

Also I prefer to miss some items rather than delete good items.